### ISSUE TITLE
Major refactoring?

### ISSUE THREAD

**Issue by alphapapa** [alphapapa/outshine#53]
@thblt @yuhan0

I wonder if we should do a survey of Outshine's features and consider which ones are still necessary to implement in this package.  

For example, outline cycling in non-Org, non-outline buffers is already handled   by outline-minor-mode, so I'm not sure that we need to implement it again in Outshine.  ISTM that, at most, we might need to implement some convenience functions on top of outline-minor-mode.

For issuing Org commands in non-Org buffers, maybe we could move to something like the `not-org` prototype library, which would relieve us of having to reimplement Outshine versions of Org commands (if not all, then at least some of them).

The bottom line, of course, is that Emacs and Org have changed a lot since Outshine was first written.  I don't have a good grasp of where its functionality fits and overlaps now, so I'm not sure where we need to go from here.

Thanks.

**Comment by yuhan0** [comment]
As far as I know, Outline-minor-mode does *not* do cycling, only the various showing and hiding commands that are much less friendly to remember than simply hitting `TAB` multiple times. 
"Convenience functions on top of outline-minor-mode" is the whole point of the `outline-magic` bits in Outshine, so I don't quite understand that part. 

I agree about removing unnecessary features - for example there are quite a few orphan functions that aren't referenced by anything else, and "hidden line cookies" seems to be quite buggy and a much better case for overlays. 

About the use-outorg family of commands, perhaps it's better for someone who regularly uses those features to comment on them - my personal use is limited to `outshine-todo`.

**Comment by yuhan0** [comment]
Also, I think the current method of piecing together and mutating a dozen global variables in `outshine-calculate-` functions and then referencing them all over the place is rather messy and could stand to be refactored for cleanliness/modularity.

**Comment by alphapapa** [comment]
> As far as I know, Outline-minor-mode does not do cycling, only the various showing and hiding commands that are much less friendly to remember than simply hitting TAB multiple times.  "Convenience functions on top of outline-minor-mode" is the whole point of the outline-magic bits in Outshine, so I don't quite understand that part.

You may be right.  What I'm trying to get at is, Outshine seems to do many things, and I don't even understand them all, and some of it seems to depend on `outorg`.  So I think it would be helpful for us if we, say, refactored its functionality into independent parts (and maybe separate files, as well).

**Comment by yuhan0** [comment]
See the most recent PR #55, which should help a little in the way of grouping independent functionality. 
Basically most of the work of separating the `outorg` commands into a separate file had already been done back in bd43f2bbb92ea0b1aedaa528a80aa065f712fccd, except it seems to have been only partially completed, leaving duplicated functions and commented regions which I just cleaned up. (reducing the size of the main `outshine.el` file down to a more manageable sounding 2500 LOC)

**Comment by alphapapa** [comment]
@thblt Looking at https://github.com/alphapapa/outorg/issues/9, and considering how closely these packages are interrelated, I'm wondering if we should make a GitHub "organization" to hold these repos.  What do you think?

**Comment by Fuco1** [comment]
I just set out to make something on top of outline which will fold/show/unfold just by hitting tab like in org-mode.  The default outline UX is horrendous... and one of the reasons org exists in the first place.

If you want to implement that I can maybe volunteer to try to hack on it if we can make it fit nicely with the rest of the package.

**Comment by tj64** [comment]
Here some historical remarks (as far as I remember):

all those **outline extensions** that went into outshine did things outline-(minor-mode) could not do, but maybe it improved in the maintime and some of these functions are obsolete duplicates now?

**hidden line cookies**
there is acually a repository (tj64/org-hlc), and I think that library was in a working state.
So all hidden line cookie functonality could be removed from outshine, anybody interested could use 
org-hlc instead. I wanted to do that (remove it) but in the end did not make it.
There is an overlay based implementation on github too, forgot the name unfortunately, from the author of poporg.el. Its very nice. But overlays have their limits with bigger files, things might get very slow.


**use outorg**
 when I wrote outshine, quite a few "org-mode" feature requests came in, and I wanted some features from org-mode too. Reusing org code if often difficult to impossible, rewriting not an option, so using outorg from outshine seemed the way to go, and it works quite well, except when permanent org files are expected (e.g. clocking). 

By script I created templates for almost all org command, implemented a few, and left the others for users interested in that functionality. If all org commands that make sense in programming mode would 
be implemented with "outshine-use-outorg", outshine headers would be just as 'smart' as org headers.

Its all prepared, but somebody would need to do the work.
And for clocking etc one would need new concepts, like a permanenet associated org file or so.

WRT dependency, oushines works without outorg (when outshine-use-outorg is omitted), but outorg needs outshine. Both libraries are a bit complex, so merging is maybe not such a good idea?
I wanted outshine-use-outorg and could not find a way to avoid mutual dependency.

**imenu**
I never used, but implemented due to feature request, and remember it worked

**other ideas** 
I remember I had some ideas started, but did not really implement, but I don't remember which unfortunately

@thblt 
"making Outshine an autonomous, generic, structural editor for various text formats."
Wow, that sounds really ambitious, but interesting too

Cheers
Thorsten

**Comment by yuhan0** [comment]
@thblt that sounds like an interesting idea for a completely different package as it would probably require a complete ground-up rewrite of the existing Outshine code. 
I think outshine's method of "hijacking" `outline-regexp` isn't an issue, it's one of those variables like `beginning-of-defun-function` that's meant to be overridden. 
It also means that we can reuse all the folding and structure manipulation functions from `outline-minor-mode`, and automatically interface with other modes which expect semantic grouping of outline structures for free. E.g. with `lispy-mode`, calling `lispy-eval` on a outline heading sends the entire subtree to the repl, and this works automatically with Outshine.

**Comment by rgrinberg** [comment]
I've switch from outshine to https://github.com/tarsius/bicycle and I'm not sure what I'm missing out. If I just need a simple mechanism for toggling outline visibility, it seems like bicycle offers the same thing at much less complexity. What am I missing out on?

**Comment by alphapapa** [comment]
@rgrinberg You can compare the packages' documentation to understand what they do and don't do.  This issue is for discussing a particular topic.  Please respect that.

**Comment by simonmichael** [comment]
Thanks for thinking about this. 3 years later, count me in favour of any such cleanup. Straightforward tab-cycling of document sections with easily customizable headings is still a valuable feature and nothing seems to do it well. (I have had it working most with outshine but it's rather difficult.)
